segmentation.py

```diff
@@ -0,0 +1,575 @@
+
+            # Update the sufficient statistics.
+            hypotheses['distribution'].update(self.__featfun(X), Y)
+
+            # Compute the log-normalization constant after the update
+            # (posterior parameter distribution).
+            # (Equation 8)
+            n_o = hypotheses['log_constant']
+            n_k = hypotheses['log_constant'] = hypotheses['distribution'].logconst()
+
+            # Evaluate the log-density of the predictive distribution.
+            # (Equation 16)
+            log_density = n_k - n_o - k * (0.5 * m * n) * np.log(2.0 * np.pi)
+
+            # Increment the counter.
+            hypotheses['count'] += 1
+
+            # Accumulate the log-likelihood of the data.
+            # (Equation 17)
+            hazard = self.__ratefun(hypotheses['count'])
+            aux = np.log(hazard) + log_density + hypotheses['log_probability']
+            loglik = self.__accum(loglik, aux)
+
+            # Keep track of the highest, log-likelihood.
+            if aux > logmax:
+                logmax, ind = aux, hypotheses['count']
+
+            # Update and accumulate the log-probabilities.
+            hypotheses['log_probability'] += np.log1p(-hazard) + log_density
+            logsum = self.__accum(logsum, hypotheses['log_probability'])
+
+        stat = MatrixVariateNormalInvGamma(self.__mu,
+                                           self.__omega,
+                                           self.__sigma,
+                                           self.__eta)
+
+        # Add a new hypothesis, which states that the next segment is about to
+        # begin.
+        self.__hypotheses.append({'index': self.__update_count,
+                                  'count': 0,
+                                  'log_probability': loglik,
+                                  'distribution': stat,
+                                  'log_constant': stat.logconst()})
+
+        logsum = self.__accum(logsum, loglik)
+        self.logsum = logsum
+
+        # Normalize the hypotheses so that their probabilities sum to one.
+        for hypothesis in self.__hypotheses:
+            hypothesis['log_probability'] -= logsum
+
+        # Automatically trim hypothesis on each update if requested.
+        if self.__maximum_hypotheses is not None:
+            self.trim(minprob=self.__minimum_probability,
+                      maxhypot=self.__maximum_hypotheses)
+
+        # In the max-product algorithm, keep track of the most likely
+        # hypotheses.
+        if self.__alg__ == 'maxprod':
+            loglik = logmax
+            self.__index.append(ind)
+
+        # In the sum-product algorithm, keep track of the probabilities.
+        else:
+            iteration = list()
+            for hypothesis in self.__hypotheses:
+                iteration.append((hypothesis['count'],
+                                  hypothesis['log_probability']))
+
+            self.__probabilities.append(iteration)
```

Comment:
hmm... this is true. An alternative might be to allow an optional limitation on the amount of data stored? A rotating buffer of N-observations where old probabilities or segments are pushed off the head? This would push the memory management back to the user but they wouldn't have to execute code externally.